libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/EnhancedLocationEngineDecorator.java

```diff
@@ -0,0 +1,185 @@
+package com.mapbox.services.android.navigation.v5.navigation;
+
+import android.app.PendingIntent;
+import android.location.Location;
+import android.os.Handler;
+import android.os.Looper;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.mapbox.android.core.location.LocationEngine;
+import com.mapbox.android.core.location.LocationEngineCallback;
+import com.mapbox.android.core.location.LocationEngineRequest;
+import com.mapbox.android.core.location.LocationEngineResult;
+import com.mapbox.geojson.Point;
+import com.mapbox.navigator.FixLocation;
+import com.mapbox.navigator.NavigationStatus;
+import com.mapbox.services.android.navigation.v5.internal.navigation.MapboxNavigator;
+
+import java.io.File;
+import java.util.Date;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import timber.log.Timber;
+
+class EnhancedLocationEngineDecorator implements LocationEngine {
+  private static final long UPDATE_INTERVAL_IN_MILLISECONDS = 1000;
+  private static final long FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS = 500;
+  private final LocationEngine locationEngine;
+  private final MapboxNavigator mapboxNavigator;
+  private final LocationEngineCallback<LocationEngineResult> callback = new CurrentLocationEngineCallback(this);
+  private ScheduledFuture future;
+  private final ScheduledExecutorService executorService;
+  private Location rawLocation;
+  private Location enhancedLocation;
+
+  EnhancedLocationEngineDecorator(@NonNull LocationEngine locationEngine,
+                                  @NonNull MapboxNavigator mapboxNavigator,
+                                  @NonNull ScheduledExecutorService executorService) {
+    this.locationEngine = locationEngine;
+    this.mapboxNavigator = mapboxNavigator;
+    this.executorService = executorService;
+    onFreeDrive();
+  }
+
+  @Override
+  public void getLastLocation(@NonNull LocationEngineCallback<LocationEngineResult> callback) throws SecurityException {
+    locationEngine.getLastLocation(callback);
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     @NonNull final LocationEngineCallback<LocationEngineResult> callback,
+                                     @Nullable Looper looper) throws SecurityException {
+    locationEngine.requestLocationUpdates(request, callback, null);
```

Comment:
should we pass `looper` instead of `null` here?

Reply:
`null` just routes to the main thread which should be fine

Reply:
Yeah, I agree.

Fixing.

Reply:
Good point. We should ensure that communication with nav-native is on a single thread only then.